### PR TITLE
Support `IGrainStorageSerializer` for memory grain storage

### PR DESCRIPTION
**Pull request by shoneefd**
A lightweight code change to implement the IStorageProviderSerializerOptions interface to MemoryGrainStorageOptions. Issue: https://github.com/dotnet/orleans/issues/8101

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/orleans/pull/8250)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8250) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8250) before we can accept your contribution.<br/><br/>:x: shoneefd [sign now](https://cla.dotnetfoundation.org/dotnet/orleans?pullRequest=8250)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/orleans?pullRequest=8250) it.</sub>

File: src/Orleans.Persistence.Memory/Storage/MemoryStorage.cs [hunk 1]
```diff
@@ -124,6 +134,53 @@ private IMemoryStorageGrain GetStorageGrain(string id)
...
+
+                logger.LogError(exc, "{Message}", sb.ToString());
+                throw new AggregateException(sb.ToString(), exc);
+            }
+
+            return dataValue;
+        }
+
+        /// <summary>
+        /// Serialize to Azure storage format in either binary or JSON format.
```
Comment (ReubenBond) [inline src/Orleans.Persistence.Memory/Storage/MemoryStorage.cs:171]:
```suggestion
        /// Serialize to the storage format.
```

File: src/Orleans.Persistence.Memory/Storage/MemoryStorage.cs [hunk 2]
```diff
@@ -87,7 +92,12 @@ public virtual async Task WriteStateAsync<T>(string grainType, GrainId grainId,
             IMemoryStorageGrain storageGrain = GetStorageGrain(key);
             try
             {
-                grainState.ETag = await storageGrain.WriteStateAsync(key, grainState);
+                var data = ConvertToStorageFormat<T>(grainState.State);
+                var binaryGrainState = new GrainState<BinaryData>(data, grainState.ETag)
```
Comment (ReubenBond) [inline src/Orleans.Persistence.Memory/Storage/MemoryStorage.cs:96]: Use `byte[]` instead of `BinaryData`

**Comment by ReubenBond** [comment]
@shoneefd thank you for this PR. I will jump in and help with the test breaks :)

**Pull request merged**